File: CombinationSum1/CombinationSum1.cpp
// CombinationSum1.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <vector>

void GetCombinations(const std::vector<int>& vec, int target, size_t index, std::vector<int>& v, std::vector<std::vector<int>>& result)
{
    if (target == 0)
    {
        result.push_back(v);
        return;
    }

    if (index == vec.size() || target < 0)
        return;

    // re-using the current element
    v.push_back(vec[index]);
    GetCombinations(vec, target - vec[index], index, v, result);
    v.pop_back();

    // use all the elements after the current one
    for (size_t i = index + 1; i < vec.size(); ++i)
    {
        v.push_back(vec[i]);
        GetCombinations(vec, target - vec[i], i, v, result);
        v.pop_back();
    }
}

std::vector<std::vector<int>> CombinationSum(std::vector<int>& A, int val)
{
    std::vector<std::vector<int>> result;
    std::vector<int> v;
    GetCombinations(A, val, 0, v, result);

    return result;
}

int main()
{
    while (true)
    {
        std::cout << "Number of elements in array: ";
        int count = 0;
        std::cin >> count;
        if (count == 0)
            break;

        std::cout << "Numbers in array: ";
        std::vector<int> vec(count, 0);
        for (int i = 0; i < count; ++i)
            std::cin >> vec[i];

        std::cout << "The sum: ";
        int sum;
        std::cin >> sum;

        std::vector<std::vector<int>> r = CombinationSum(vec, sum);
        for (auto& vv : r)
        {
            if (!vv.empty())
                std::cout << vv[0];
            for (auto i = vv.begin() + 1; i != vv.end(); ++i)
                std::cout << ", " << *i;
            std::cout << std::endl;
        }
    }
    std::cout << std::endl;
}

File: Go/CombinationSum1.go
package main

import "fmt"

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetCombination(A []int, val, index int, v []int, result *[][]int) {
	if val == 0 {
		c := make([]int, len(v))
		copy(c, v)
		*result = append(*result, c)
		return
	}

	if index == len(A) || val < 0 {
		return
	}

	for i := index; i < len(A); i++ {
		v = append(v, A[i])
		GetCombination(A, val-A[i], i, v, result)
		v = v[:len(v)-1]
	}
}

func CombinationSum(A []int, val int) [][]int {
	var result [][]int
	var r []int

	GetCombination(A, val, 0, r, &result)

	return result
}

func main() {
	for {
		fmt.Print("Number of elements: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := GetArrayFromInput(count, "Please enter the array: ")

		fmt.Print("Please enter the sum: ")
		var val int
		fmt.Scan(&val)

		result := CombinationSum(vec, val)

		fmt.Println("The combinations that have a sum of ", val, ": ", result)
	}
}
